Support jwt-bearer grant-type (RFC7523) in oauth2 extension

JorTurFer opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
This PR supports the grant-type defined by [RFC7523](https://datatracker.ietf.org/doc/html/rfc7523) "JSON Web Token (JWT) Profile for OAuth 2.0 Client Authentication and Authorization Grants" in extensiont/oauth2clientauth.

This flow is needed for some OAuth providers, like [STACKIT Service Accounts](https://docs.stackit.cloud/stackit/en/usage-of-the-service-account-keys-in-stackit-175112464.html)


<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/44314

<!--Describe what testing was performed and which tests were added.-->
#### Testing
I've added a test where the signed token is evaluated field by field. I've also test locally using STACKIT Service Account endpoint and I've retrieved a valid access token. 


<!--Describe the documentation added.-->
#### Documentation

<!--Please delete paragraphs that you did not use before submitting.-->

JorTurFer committed:
Support jwt-bearer grant-type (RFC7524) in oatuh2 extension

Signed-off-by: Jorge Turrado <jorge.turrado@mail.schwarz>

atoulme commented on extension/oauth2clientauthextension/config.go:
```diff
@@ -35,6 +37,43 @@ type Config struct {
 	ClientSecretFile string `mapstructure:"client_secret_file"`
 
+	// ClientCertificateKeyID is the Key ID to include in the jwt. Only used if
+	// GrantType is set to "urn:ietf:params:oauth:grant-type:jwt-bearer".
```
do we error out if ClientCertificateKeyID is set and GrantType is not "urn:ietf:params:oauth:grant-type:jwt-bearer" ?

atoulme replied:
It's likely we can leave it as it is for a while. Let's see when people bump into this and complain they expect a different behavior.

atoulme commented on extension/oauth2clientauthextension/config.go:
```diff
@@ -35,6 +37,43 @@ type Config struct {
+
+	// ClientCertificateKey is the application's private key. Only used if
+	// GrantType is set to "urn:ietf:params:oauth:grant-type:jwt-bearer".
+	ClientCertificateKey configopaque.String `mapstructure:"client_certificate_key"`
```
can all those new parameters be defined in the README.md?

JorTurFer replied:
yeah for sure, I forgot to update it with an example, sorry

JorTurFer committed:
apply feedback

Signed-off-by: Jorge Turrado <jorge.turrado@mail.schwarz>

JorTurFer committed:
Add changelog

Signed-off-by: Jorge Turrado <jorge.turrado@mail.schwarz>

JorTurFer commented:
> Please take a look at the CI, add a changelog and review my comments and mark the PR ready to review again.

I've addresses all the feedback except the `goleak` one, could you trigger CI checks again please?

JorTurFer committed:
fix changelog

Signed-off-by: Jorge Turrado <jorge_turrado@hotmail.es>

JorTurFer committed:
fix codegen

Signed-off-by: Jorge Turrado <jorge_turrado@hotmail.es>

JorTurFer commented:
@atoulme, PR is ready to review 😄 
The only topic left is this -> https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44318#discussion_r2545321082

pavankrish123 reviewed (approved):
LGTM - except if we can avoid the goleak ignore in the tests

JorTurFer committed:
update goleak

Signed-off-by: Jorge Turrado <jorge.turrado@mail.schwarz>

JorTurFer committed:
readd goleak exception

Signed-off-by: Jorge Turrado <jorge.turrado@mail.schwarz>

JorTurFer commented:
This is not stale

JorTurFer committed:
tidy

Signed-off-by: Jorge Turrado <jorge.turrado@mail.schwarz>

JorTurFer committed:
lint

Signed-off-by: Jorge Turrado <jorge.turrado@mail.schwarz>

atoulme commented on extension/oauth2clientauthextension/metadata.yaml:
```diff
@@ -12,3 +12,8 @@ status:
+    ignore:
+      top:
+        - "net/http.(*persistConn).readLoop"
+        - "net/http.(*persistConn).writeLoop"
```
We can leave this for a subsequent PR. It's ok to have this for now, if it's just a test issue.

atoulme reviewed (approved):
Approved by codeowner - there's a follow up on closing connections we need to address separately, I think. LGTM and merging.
